src/out_of_order_engine.cc

```diff
@@ -0,0 +1,430 @@
+	assert(node_it != incomplete_instructions.end());
+	auto deleted_node = std::move(node_it->second); // move so we can access members / iterate successors after erasure
+	incomplete_instructions.erase(node_it);
+
+	auto& was_assigned = std::get<assigned_state>(deleted_node.assignment);
+	if(deleted_node.target == target::host_queue || deleted_node.target == target::device_queue) {
+		// "remove" instruction from assigned lane
+		assert(was_assigned.lane.has_value());
+		auto& lane = get_lane_state(deleted_node.target, was_assigned.device, *was_assigned.lane);
+		assert(lane.num_in_flight_assigned_instructions > 0);
+		lane.num_in_flight_assigned_instructions -= 1;
+		if(lane.last_incomplete_submission == instr) { lane.last_incomplete_submission = nullptr; }
+	}
+
+	for(const auto succ_iid : deleted_node.successors) {
+		// find() instead of at(): successor might have completed out-of-order
+		if(const auto succ_it = incomplete_instructions.find(succ_iid); succ_it != incomplete_instructions.end()) {
+			auto& successor = succ_it->second;
+			assert(successor.num_incomplete_predecessors > 0);
+			--successor.num_incomplete_predecessors;
+			try_mark_for_assignment(successor);
+		}
+	}
+}
+
+bool engine_impl::is_idle() const {
+#ifndef NDEBUG
+	if(incomplete_instructions.empty()) {
+		assert(assignment_queue.empty());
+		for(const auto& device_state : device_queue_target_states) {
+			for(const auto& lane : device_state.lanes) {
+				assert(lane.num_in_flight_assigned_instructions == 0);
+			}
+		}
+		for(const auto& lane : host_queue_target_state.lanes) {
+			assert(lane.num_in_flight_assigned_instructions == 0);
+		}
+	}
+#endif
+	return incomplete_instructions.empty();
+}
+
+incomplete_instruction_state* engine_impl::pop_assignable() {
+	while(!assignment_queue.empty()) {
+		const auto instr = assignment_queue.top();
+		assignment_queue.pop();
+
+		auto& node = incomplete_instructions.at(instr->get_id());
+		assert(node.num_unassigned_predecessors == 0);
+
+		if(const auto eagerly_assignable_when_pushed = std::get_if<conditional_eagerly_assignable_state>(&node.assignment)) {
+			assert(node.num_incomplete_predecessors > 0); // otherwise this would be an immediately_assignable_state
+			const auto& lane = get_lane_state(node.target, eagerly_assignable_when_pushed->device, eagerly_assignable_when_pushed->lane);
+			if(lane.last_incomplete_submission == nullptr
+			    || lane.last_incomplete_submission == eagerly_assignable_when_pushed->expected_last_submission_on_lane) {
+				// Our preferred lane is still in the required state to go through with eager assignment
+				return &node;
+			} else {
+				// A third instruction has been submitted to our preferred lane since, so we drop eager assignment and don't need to attempt it again, since
+				// none of our dependencies can now end up last in the queue anymore.
+				node.assignment.emplace<unassigned_state>().probe_for_eager_assignment = false;
```

Comment:
That's what we hope for, exactly. Going through with eager assignment in this case means we would serialize instructions that can otherwise be concurrent.